packages/api/src/objectSet/ObjectSet.ts

```diff
@@ -142,9 +144,48 @@ export interface InterfaceObjectSet<
 > extends MinimalObjectSet<Q> {
 }
 
+export type ExtractRdp<
+  D extends ObjectSet<any, any> | Record<string, SimplePropertyDef>,
+> = D extends Record<string, SimplePropertyDef> ? D : {};
+
+export namespace ObjectSet {
+  export namespace Fns {
+    export interface WithProperties<
+      Q extends ObjectOrInterfaceDefinition = any,
+      D extends Record<string, SimplePropertyDef> = {},
+    > {
+      readonly withProperties: <
+        R extends Record<string, SimplePropertyDef>,
+      >(
+        clause: { [K in keyof R]: Rdp.Selector<Q, R[K]> },
+      ) => ObjectSet<
+        Q,
+        {
+          [NN in keyof R | keyof D]: NN extends keyof R ? R[NN]
+            : NN extends keyof D ? D[NN]
+            : never;
+        }
+      >;
+    }
+  }
+}
+
 export interface ObjectSet<
   Q extends ObjectOrInterfaceDefinition = any,
-  _UNUSED = any,
+  // Generated code has what is basically ObjectSet<Q> set in here
+  // but we never used it so I am repurposing it for RDP
+  D extends ObjectSet<Q, any> | Record<string, SimplePropertyDef> = ObjectSet<
+    Q,
+    any
+  >,
+> extends
+  ObjectSetPrime<MergeObjectSet<Q, D>>,
+  ObjectSet.Fns.WithProperties<Q, ExtractRdp<D>>
+{
+}
+
+interface ObjectSetPrime<
+  Q extends ObjectOrInterfaceDefinition = any,
```

Comment:
Splitting meat out so we can run `MergeObjectSet<Q,D>` once instead of every place there was a `Q`.